Let user update the app

kekkyojin opened the pull request:
## PR Checklist

<!-- For the checkbox formatting to work properly, make sure there are no spaces on either side of the "x" -->

Please check all that apply to this PR using "x":

- [x] I have checked that this PR is not a duplicate of an existing PR (open, closed or merged)
- [x] I have checked that this PR does not introduce a breaking change

## PR Type

What kind of change does this PR introduce?

- [x] Feature

## Fixes

Issue Number:
 #328

## What is the current behavior?

For packages installed not from Play Store or FDroid, user needs to manually download the app and install it
## What is the new behavior?

The app will show a snackbar when a new version is released, allowing user to install it

## Notes
This PR requires that certain file exists on the update server. That file must be have a JSON format. The specification is documented on this PR code.

JSON file name and hosting can be changed. In order for this to be tested, I used my own web server, where I put a JSON file and an APK built from master branch but with a version code of 17.

To avoid accidental updates of the app while developing, the code will not run for DEBUG builds. In order to test this PR, you will need to comment out the corresponding code which checks for DEBUG builds.

New version code checking will run once every 24 hours, but user could still force the checking by quiting the app and restarting it.

kekkyojin committed:
Refactor comment child items visibility switching code (#168)

kekkyojin committed:
Fix CreditsBalanceView position on channel creation dialogs (#171)

kcinyiama committed:
- Fix crash caused when sorting comments with no reactions (#177)

akinwale committed:
Library rework (#179)

* rework Library view to display recently watched and playlists

* load public playlists

* do not save collections to view history

akinwale committed:
Merge pull request #178 from kossy18/master

- Fix crash caused when sorting comments with no reactions

akinwale committed:
save view history for claim in collection

kcinyiama committed:
- Added null checks when retrieving the wallet available and total balances.
- Displayed the FileViewFragment also if a video is playing when resumed (if pip mode was switched off and on).
- Prevented the wallet balance details from being if the wallet balance object is null especially when there's no data connection.
- Added a null check when clearing notifications which could happen during a sign out.
- Fetched channels at delayed intervals when there's no data connection

ktprograms committed:
Add FOSS build variant (#176)

ktprograms committed:
Add GitHub Actions to build Full and Foss APKs

ktprograms committed:
Handle odysee.com links

akinwale committed:
Merge pull request #181 from kossy18/master

Fixed some issues

akinwale committed:
Merge pull request #185 from ktprograms/gh-actions

Add GitHub Actions to build Full and Foss APKs

akinwale committed:
Merge pull request #187 from OdyseeTeam/odysee-open-link

Handle odysee.com links

kekkyojin committed:
Use ExoPlayer instead of deprecated SimpleExoPlayer (#183)

kekkyojin committed:
Fix Discover link (#184)

ktprograms committed:
Implement direct livestreaming (#167)

* Implement direct livestreaming (Extremely WIP)

* Move toHexString to Helper and reformat if statement

* Suspend/resume global player when entering/exiting GoLiveFragment

* Use rationale string customized for livestreaming

* Move checking for camera permission into onResume

This fixes the camera permission prompt not being shown after the audio
permission prompt is shown.

* Remove commons-lang3 dependency

It was removed in HaishinKit.kt commit 51140995a5.

* Remove extra tools attributes, add newline to end of fragment_go_live

* Add progress indicator when publishing livestream

Also disable the "Start Streaming" button while publishing.

* Add thumbnail uploading and set livestream thumbnail

Also properly add and remove listeners, use shouldSuspendGlobalPlayer
instead of manually suspending and resuming the global player.

* Set Claim meta from input JSON

Used for getting the claim's effectiveAmount.

* Add precheck and error handling

Logic mostly copied verbatim from the iOS code.

* Add tools:text to channel error text, use padding instead of margin

Using padding instead of margin is recommended by Android Studio for
RTL symmetry.

* Keep screen on while livestreaming

* Automatically stop stream and go to home screen when waking

Livestreaming sometimes doesn't resume properly so just stop the stream.

* Remove explicit false assignment to state variables

* Add confirmation dialog when back pressed while streaming

* Update HaishinKit dependency

There is a partial fix to the stretching bug

* Refactor to put livestreaming in an Activity

* Use ClaimListSupplier on API 24+

* Wait for stream claim to be confirmed before starting

* Use METHOD_TXO_LIST instead of hardcoded string

* Reorder PublishClaimTask member variables to match parameters

* Make authToken final in PublishClaimTask

* Use transparent black for switch camera button tint

* Call super.onBackPressed() when not streaming

Fix not being able to go back while setting up stream or waiting for
confirmation.

* Close cameraSource in onPause, allow switching apps while confirming

Do not call fetchChannels (which later shows the livestream set up
screen) if resuming activity while waiting for confirmation.

* Update HaishinKit, add configChanges to GoLiveActivity

- HaishinKit commit 018d64a8883d66b50fcc49631de4cc336ad756eb fixed
  the stream being stretched when setting resolution.
- Make GoLiveActivity not restart when screen rotated.

kekkyojin committed:
Show LBRY Credits icon on CreditsBalanceView at design-time (#191)

kekkyojin committed:
(Regression) Show the clear all button on LibraryFragment (#197)

kekkyojin committed:
Fix status bar foreground color for Android 12+ (#199)

kekkyojin committed:
Show snackbars at the bottom (#196)

kekkyojin committed:
Use correct Content-Type for HLS content (#200)

kekkyojin committed:
Simplify AllContent layout and increase interactive area (#198)

akinwale committed:
implement livestream chat

akinwale committed:
Merge pull request #203 from OdyseeTeam/livestream-chat

implement livestream chat

kekkyojin committed:
Use Livestream API queries on content, Following and Channel fragments (#189)

ktprograms committed:
Enable publishing (#182)

* Enable publishing

Add TusPublishTask to upload file using Tus endpoint and add claim info.

* Re-add limit 150 to gallery load

With split to use new method for Android 11+

* Re-enable camera preview in publish fragment

Update CameraX libraries to 1.1.0-beta03

* Publish: Remove photo button, set file picker to video only

ktprograms committed:
Show images and markdown posts (#202)

* Use total_pages for checking if end reached in claim search

Fix a bug where hasReachedEnd was incorrectly set because the claims
array was smaller than page_size due to filtering.

* Following: Start search page at 1

LBRY page index starts at 1, there is no result at page 0.

* Search: When filter opened, only show type picker if file chip selected

* Display markdown posts and images

- Use org.commonmark dependency and add extensions dependencies.
- Allow filtering for images and text in search.
- Put search file type list in scroll view.

Fix: #186

ktprograms committed:
Show playback progress indicator on thumbnail (#201)

* Save and load claim duration in view history

* Show playback progress indicator on thumbnail

kekkyojin committed:
Scroll to comment from notification item (#192)

kekkyojin committed:
Add livestream chat messages from UI thread (#206)

kekkyojin committed:
Rename childsToBeShown instance field to childrenToBeShown (#207)

kekkyojin committed:
Ensure claim object is not null before processing (#205)

ktprograms committed:
Remove gh-actions branch trigger from CI workflow

Branch was used for testing and no longer exists.

akinwale committed:
Merge pull request #208 from OdyseeTeam/remove-workflow-ghaction-branch

Remove gh-actions branch trigger from CI workflow

ktprograms committed:
Fix media controls notification showing disambiguation drawer

Use direct intent to MainActivity instead of deep link intent.

Use non-deprecated PlayerNotificationManager.Builder constructor.

Fix: #210

ktprograms committed:
Fix wrong playback default speed preference key

Fix: #211

akinwale committed:
Merge pull request #212 from OdyseeTeam/media-controls-notification-no-disambiguation

Fix media controls notification showing disambiguation drawer

akinwale committed:
Merge pull request #213 from OdyseeTeam/default-playback-speed-fix-settings-key

Fix wrong playback default speed preference key

kekkyojin committed:
Use smooth scrolling when adding new chat messages on livestreams (#214)

kekkyojin committed:
Increase timeout amount for Lighthouse search queries (#209)

kekkyojin committed:
Decrease number of network requests while in Battery Saver mode (#221)

kekkyojin committed:
Show appropriate UI when a claim is for a non-live livestream (#220)

kekkyojin committed:
Show Comments tab on Channel fragment (#223)

ktprograms committed:
Animation when player minimized/maximized

- Keep main_activity_other_fragment always visible so fragment is still
  visible during transaction. This is fine as the fragment is removed at
  the end of the transaction.
- Vertically stack fragment containers so main_activity_other_fragment
  can be left visible. The main_activity_other_fragment container is
  placed top-most so the animation is always visible.
- Use ?android:attr/colorBackground in FileViewFragment so it is not
  transparent during the exit transition.

akinwale committed:
Signed builds with Github Actions (#226)

* add build-release job for signed APKs

* import PGP key in CI workflow

* update apk output path in release build script

* fix apk output path

* post-build tweak

kekkyojin committed:
Show No Comments uI when appropiate (#224)

kekkyojin committed:
Some nullity checks (#225)

ktprograms committed:
File View: Detect downwards fling to dismiss

kekkyojin committed:
Determine if video was transcoded based on the response headers (#227)

kekkyojin committed:
Replace transaction list async task with a callable (#228)

ktprograms committed:
Change magic link sign in text

ktprograms committed:
Following notifications: default off, show confirm change alert

akinwale committed:
drive sign in when the notifications icon is tapped

akinwale committed:
drive sign up / sign in before allowing publish

akinwale committed:
do not display mature tags for publishing

ktprograms committed:
Bump HaishinKit.kt to 0.9.4

Use upstream version as it has JitPack build support.

Bump compile and target SDK version to 31 to be compatible with
HaishinKit.kt 0.9.4.

ktprograms committed:
Fix dark mode icon color in Library

ktprograms committed:
Add Menu button to show Publishes, use Auth Token in Claim List

Add ClaimListTask constructor that takes list of types and auth token.

kekkyojin committed:
Use System default theme as default dark mode setting (#233)

ktprograms committed:
Channel content: Set contentHasReachedEnd based on Page isLastPage

- Refactor claim list callables to return Page
- Individually parse result of each callable to get isLastPage value
  from last callable (claim search).

ktprograms committed:
Use channel title on file claim preview

ktprograms committed:
Set playlist title when playing from non-local collections

Fix "null • x/x" being shown when playing from non-local collections.

kekkyojin committed:
Fix global player not expanding to full content when clicked (#240)

kekkyojin committed:
Only show Scheduled Livestreams tab for a channel when appropiate (#242)

kekkyojin committed:
Use global Executors (#243)

ktprograms committed:
Add pagination to claim list

- Move claimList to method in Lbry.
- Use Page instead of List<Claim> in ClaimListTask.
- Create new ResolveResultHandler that doesn't use page (previously
  ResolveTask was using ClaimListResultHandler).
- Remove small loading view from Publishes fragment, just use big
  loading (reanamed to loading) everywhere.

ktprograms committed:
Show publisher thumbnail in claim preview

Move publish time below publisher (same as on odysee.com)

ktprograms committed:
Show view count on claim preview (only in channel page)

ktprograms committed:
Hide claims view count view when getViews() returns null

i.e. on all places other than channel content, or when error occurred.

- Also hide in channel content if there's zero views
- Show "No views" instead of "0 views" in FileViewFragment

ktprograms committed:
Fix playback restarting when opening claim from global player

- Save canonicalUrl in claimCache.
- Cache claim even when not resolving a repost.
- Add FLAG_IMMUTABLE to createCurrentContentIntent to fix crash on
  Android 12 when clicking global player.

ktprograms committed:
Rearrange layout of player controls

- Move Quality and Speed selectors to top-right
- Hide Quality selector when only Auto is available
- Use bold text for Quality and Speed selectors
- Make duration views wrap_content width so progress bar can be wider

ktprograms committed:
Set Livestream claim release time to start time from API

ktprograms committed:
Only show "LIVE" for total duration on livestreams

Elapsed duration is still shown as it works properly.

ktprograms committed:
Following: Fix livestreams not being shown when filtering by channel

kekkyojin committed:
Use Application scheduled executor on FileViewFragment (#255)

kekkyojin committed:
Drive user to sign in to react (#256)

ktprograms committed:
Save and switch back to previously shown fragment in channel page (#247)

ktprograms committed:
Set playback rate to 1x when livestream has reached live edge

- When buffer event happens, get player timeline window and compare
  playback position with default (close to edge) position. If current
  playback position is newer than default, set playback rate to 1x.
- Use setPlaybackSpeed() instead of setPlaybackParameters().

Note that this _can_ cause playback rate to reset if skipping to an area
that not loaded yet _and_ in front of the live edge. It shouldn't be a
major problem as when just watching at 2x to catch up on a livestream it
wouldn't go past the buffered area.

ktprograms committed:
Stop video autoplaying/stopping when switching to another claim

Add play icon (used instead of main action button) for manually starting
videos.

ktprograms committed:
Hardcode videoIsTranscoded = true on livestreams

Fix quality selector not being shown on livestreams.

akinwale committed:
fix notification author URLs for resolve

kekkyojin committed:
Disable Google Cast dependency while not implemented (#257)

akinwale committed:
Fix sign in flow initiated from SignInActivity (#263)

* add debug logs

* make dbHelper reference static

* allCollections null check

ktprograms committed:
Don't include list related context menu items on playlists

ktprograms committed:
Hide play/pause when buffering

ktprograms committed:
Change Autoplay setting to 3-way list

Option for "Always", "Never", or "Only when nothing is playing".

ktprograms committed:
Hide player surface while switching videos

Fix having a frame from the previous video shown while new video is
loading.

kekkyojin committed:
Use VERSION_CODES.S (#266)

ktprograms committed:
Show publish time as relative always, except on FileViewFragment

- Add PrettyTime library (also used by NewPipe).
- Create FormatTime helper class.
- Use DateTimeFormatter to show absolute localized time in
  FileViewFragment. There should be at least one place where the
  absolute publish time is visible.
- Parse notification creation time with Instant.parse().

ktprograms committed:
Fix crash getting thumbnail URL for null signing channel

ktprograms committed:
Open Community Guidelines and Help & Support links in-app

- Add call to openFragment to openChannelUrl method.

ktprograms committed:
Fix tapping on media controls notification not opening player

Use nowPlayingClaim.getCanonicalUrl() if nowPlayingClaimUrl is null.

ktprograms committed:
Use compact formatting for channel content claim view counts

ktprograms committed:
Fix playback speed resetting when jumping around

Only use timeline window edge logic when playing a livestream.

ktprograms committed:
Fix default speed and quality ignored on claims from openFileClaim()

openFileClaim() sets "claim" param, so the `newClaim == null` code path,
which calls onNewClaim(), which sets `loadingNewClaim`, is not called.

ktprograms committed:
Reorder attributes in settings.xml

ktprograms committed:
Fix infinite recursion on ChannelCreateDialog showError method

Call showError on MainActivity and dismiss dialog instead.

ktprograms committed:
Use SimpleMenuPreference library for better dark mode contrast

akinwale committed:
prevent crash if builtinCollections is not present in wallet state

akinwale committed:
unpublishedCollections from wallet state null check

kekkyojin committed:
Use Collectors.joining() on up to API Level 25 (#267)

kekkyojin committed:
Don't URL encode query on Lighthouse (#268)

kekkyojin committed:
Make player UI changes from main thread instead of worker one (#282)

akinwale committed:
Rewards cleanup (#281)

* tweak reward list item display and text sizes

* add reward verification flow to Rewards screen

* display summary for identity_verified=True but reward_approved=False

kekkyojin committed:
Use start/stop time from chat websocket (#276)

kekkyojin committed:
Show publish time for broadcast which is not active

akinwale committed:
set style property in ClaimListAdapter constructor

akinwale committed:
fix layout for recently watched items

ktprograms committed:
Use semi-transparent Odysee pink for featured claims

Adds contrast to featured claims in dark mode.

Use ?android:attr/textColorPrimary for text.

ktprograms committed:
Improve resolving of featured (matching URL) claims

- Trim leading/trailing whitespace and replace with hyphen between
  words.
- Resolve channel claim from query as well.
- Resolve (and add to list) featured claims only after initial search
  completed.
- Hide duration view when showing featured channel claim.
- Hide other info views when showing "Publish something" item
  (when query does not resolve).
- Remove duplicate percent sign and redundant escape in invalid URI
  regex.

ktprograms committed:
Lighthouse: Remove resolve=true, run resolve after search

Filter null results and subtract from expected page size (so pagination
works correctly).

Use Ordering.explicit to put claims in correct order after resolve.

ktprograms committed:
Show original claim short URL for reposted featured claims

ktprograms committed:
Implement removing from lists

ktprograms committed:
File view: load claims in pages of 50

Use Ordering.explicit for sorting instead of Map technique.

ktprograms committed:
Run Unit Tests in CI

tzarebczan committed:
Merge pull request #285 from OdyseeTeam/no-broadcasting-date

Show publish time for non-active livestream claim

kekkyojin committed:
Allow to set a default channel used to interact with the platform (#283)

kekkyojin committed:
Use odycdn CDN and change image files sizes (#241)

Co-authored-by: Thomas Zarebczan <thomas.zarebczan@gmail.com>

kekkyojin committed:
Fix some livestreams being filtered from the channel content (#290)

ktprograms committed:
ContentSources v2 API, Wild West excluded channels

- Use ContentSources v2 API.
- Use Wild West category from API including excludedChannelIds array.
- Refresh cache when locale changes.

Fix: #195

ktprograms committed:
File view: Only show 25 items from playlist at once

Improve performance due to NestedScrollView preventing RecyclerView from
recycling (it takes maximum height so all claims are rendered).

akinwale committed:
Livestream chat improvements (#286)

* make chat member names clickable in live chat

* highlight streamer name in chat

* display emotes and stickers

* fix emote / sticker parsing

* display hyperchats

* tweak hyperchat container padding

* enable credits tipping in live chat

* do not autoscroll to the end of chat messages if not currently at the bottom of the scrollview

ktprograms committed:
Search: Add Sort By option, use Lighthouse query params

Remove filter functionality from ClaimListAdapter, use Lighthouse query
params instead.

ktprograms committed:
Update Like/Dislike count/color immediately when reacting

Reset to original values if react fails.

akinwale committed:
load and check custom blocking rules

akinwale committed:
update thumbnail upload endpoint

ktprograms committed:
Bump ExoPlayer to 2.18.1

Required for fix in cast extension 2.17.0 to depend on up-to-date cast
library that creates PendingIntent correctly.

ktprograms committed:
CI: Install NDK version 21.0.6113669

akinwale committed:
update form-data field names and response handling

kekkyojin committed:
Use Application object executors for verification email fragments (#301)

ktprograms committed:
Chromecast support

Create PlayerManager and CastHelper classes to handle cast playback as
well as split code for FOSS variant.

PlayerManger initializes players and has methods for changing current
player and initializing player with media URL.

CastHelper holds the CastContext and handles cast button visibility and
the Chromecast introductory overlay.

Remove references to Players in MainActivity and FileViewFragment.
Players are always accessed through the PlayerManager instance.

MediaRouteButton is present in the layout files in the main source set,
since it is not part of a closed source library. Its visibility remains
GONE the entire time in the FOSS variant.

Fix: #148

kekkyojin committed:
Fix NullPointerException crash when sorting related content claims (#304)

akinwale committed:
load and save creator settings

kekkyojin committed:
Allow user to hide chat messages on a livestream (#300)

akinwale committed:
load channel settings and check against user actions

ktprograms committed:
Open Publish Form when receiving video file.

Fix: #316

ktprograms committed:
Match Publish Form UX to odysee.com

- Make description field multiline.
- Remove generate random URL button.
- Indicate channel that content will be published as in URL.
- Load known tags from Predefined.
- Add release time setting.
- Remove setting claim release time from metadata. In claim editing PR
  release time will be loaded from metadata on fragment initialize.

ktprograms committed:
Fix crash creating channel in publish form

Also make channel creation dialog name field single line.

akinwale committed:
restore take photo button on publish screen

ktprograms committed:
Fix app bar titles not showing, format fragment_publish_form.xml

- Add TextView to custom app bar to display title. Built in title view
  is placed below custom views so is not visible.
- Add regions to fragment_publish_form.xml to make sections of the form
  easier to separate.
- Replace NestedScrollView -> ScrollView as it is not needed (the only
  scrolling view, the tags view, is limited to 5 added + 8 suggested,
  which doesn't need to scroll).
- Remove redundant LinearLayout from around Additional Options card

ktprograms committed:
Implement Claim edit and deletion.

- Rename Unpublish button to Delete, use Delete button's icon.
- Remove local file delete task as there's no local LBRY SDK.
- Fix editing Anonymous claims showing create channel dialog.
- Remove filePath parameter from PublishClaimTask.

ktprograms committed:
gradle: bump compileSdkVersion and targetSdkVersion to 33

ktprograms committed:
Make it possible to delete pending claims

Hide edit action if pending claim selected. PublishForm doesn't seem to
handle getting the channel for a pending claim, so just disable edit.

ktprograms committed:
Clear action bar title in Publishes and Channel Manager

Action bar title gets set by PublishForm and ChannelForm and stays set
when navigating back.

akinwale committed:
apply dark mode for sign up / sign in activity

ktprograms committed:
Fix time zone issues in Publish Form release date picker

Use UTC_ZONE instead of repeating ZoneId.of("UTC") everywhere.

akinwale committed:
fix null pointer exception

akinwale committed:
publish playlists, allow published playlists to be edited / updated

akinwale committed:
extend claimNameExists to also check in collections for existing claim names

kekkyojin committed:
Remove informational logging message (#334)

kekkyojin committed:
Show LIVE as text on list item when no viewers are on livestream (#336)

kekkyojin committed:
Replace wildwest with rabbithole as category name (#335)

kekkyojin committed:
Future scheduled livestreams (#333)

akinwale committed:
Commentron blocking (#338)

* rename block --> mute

* process commentron moderation.Block and moderation.UnBlock requests

kekkyojin committed:
Show currently active livestreams for each category on the Home tab (#341)

kekkyojin committed:
Bump Glide to 4.13.2 (#342)

kekkyojin committed:
Fix PlaylistStuckException when playing livestreams (#343)

kekkyojin committed:
Filter scheduled livestreams on Following according to channel selection (#344)

ktprograms committed:
Support dark theme for Markdown posts

ktprograms committed:
publish: fix using currency "lbry credits" when setting fee

tzarebczan committed:
Merge pull request #346 from OdyseeTeam/fee-currency-fix

publish: fix using currency "lbry credits" when setting fee

tzarebczan committed:
Merge pull request #345 from OdyseeTeam/text-post-color-scheme

Support dark theme for Markdown posts

kekkyojin committed:
Self-update app

ktprograms committed:
Match Livestreaming UX flow to odysee.com

UX flow:
- Clicking the Upload button shows menu to pick Upload or Go Live
- Create livestreams with form like publishing
- Choose New Livestream or Upload Replay
  Replays shown in pages of 4 like on odysee.com
- Choose Anytime or Scheduled Time stream
- Clicking the Create button goes to livestreams dashboard (also
  accessible from user menu)
- Livestreams dashboard shows stream server and stream key, with button
  to copy
- Dashboard waits for created claim to finish confirming then enables
  Start Streaming button
- Clicking Start Streaming goes to activity that streams with HaishinKit
- Streaming activity sets FLAG_KEEP_SCREEN_ON in onCreate instead of
  when streaming starts. Also removes flag only onDestroy.

Changes of note:
- There are now 3 publish tasks:
  - PublishClaimTask uses the LBRY API
  - ReplayPublishTask uses the v1 Odysee Publish API
  - TusPublishTask uses the v2 Odysee Publish API (uses Tus for upload)
- Add buildPublishOptions function in Helper to create options map from
  Claim.

Fix: #93
Fix: #94
Fix (partial): #232 (app reloading after accepting permissions to
camera/audio when trying to go live)

ktprograms committed:
Publish Form: More specific title for edit, refactor

Fix warnings with deposit amount code.

ktprograms committed:
Pass token to ClaimListTask when fetching own claims

Fix Lbry.ownClaims being null because ClaimListTask failed without token

ktprograms committed:
publishes: add contentDescription to floating button

akinwale committed:
add screen for managing blocked and muted channels

kekkyojin committed:
Sync default channel name selection (#347)

* Use default channel name from wallet settings data

* Save default channel name to wallet settings data

akinwale committed:
Merge pull request #349 from OdyseeTeam/livestream-improvements

Match Livestreaming UX flow to odysee.com

ktprograms committed:
Fix crash connecting to livestream chat from publishes list

If getSigningChannel().getCanonicalUrl() is null, resolve the signing
channel first.

Also set channel_id on new comments from WebSocket so streamer comments
get highlighted.

ktprograms committed:
issue_template: use correct chat link

ktprograms commented on app/src/main/java/com/odysee/app/ui/findcontent/AllContentFragment.java:
```diff
@@ -466,6 +496,172 @@ public void onPause() {
+                    if (Build.VERSION.SDK_INT > Build.VERSION_CODES.S_V2) {
+                        resInfoList = context.getPackageManager().queryIntentActivities(intent, PackageManager.ResolveInfoFlags.of(PackageManager.MATCH_DEFAULT_ONLY));
+                    } else {
+                        //noinspection deprecation
```
Android Studio says this is a "redundant suppression", and the fixit (when pressing alt+enter) removes this line.

kekkyojin replied:
Same as above

ktprograms commented on app/src/main/java/com/odysee/app/ui/findcontent/AllContentFragment.java:
```diff
@@ -466,6 +496,172 @@ public void onPause() {
+                        //noinspection deprecation
+                        resInfoList = context.getPackageManager().queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
+                    }
+                    for (ResolveInfo resolveInfo : resInfoList) {
+                        context.grantUriPermission(context.getApplicationContext().getPackageName() + ".provider", downloaded_apk, Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
+                    }
```
`resolveInfo` isn't used inside this loop. What's it meant to do?

ktprograms commented on app/src/main/res/values/strings.xml:
```diff
@@ -3,6 +3,11 @@
     <string name="navigation_drawer_close">Close navigation drawer</string>
 
+    <!-- App updater -->
+    <string name="downloadmanager_new_version_available">New version available for Odysee Android</string>
+    <string name="snackbar_new_version_available">There is a new version available for Odysee Android</string>
```
I think this wording is better:
```suggestion
    <string name="downloadmanager_downloading_new_version">Downloading new version of Odysee Android</string>
    <string name="snackbar_new_version_available">There is a new version available of Odysee Android</string>
```

ktprograms commented on app/src/main/java/com/odysee/app/ui/findcontent/AllContentFragment.java:
```diff
@@ -466,6 +496,172 @@ public void onPause() {
+                                        public void onClick(View view) {
+                                            Context c = getContext();
+                                            if (c != null) {
+                                                ContextCompat.checkSelfPermission(c, Manifest.permission.WRITE_EXTERNAL_STORAGE);
+                                                String filePath = c.getExternalFilesDir(null).getAbsolutePath().concat("/odysee-android.apk");
+                                                final Uri uri = Uri.parse("file://" + filePath);
+                                                File f = new File(filePath);
+
+                                                if (f.exists()) {
+                                                    f.delete();
+                                                }
+                                                DownloadManager.Request request = new DownloadManager.Request(Uri.parse(latestUrl));
+                                                request.setDescription("Odysee new version available");
+                                                request.setTitle(getResources().getString(R.string.downloadmanager_new_version_available));
+
+                                                //set destination
+                                                request.setDestinationUri(uri);
+                                                request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE);
+
+                                                // get download service and enqueue file
+                                                final DownloadManager manager = (DownloadManager) c.getSystemService(Context.DOWNLOAD_SERVICE);
+                                                manager.enqueue(request);
+
+                                                //set BroadcastReceiver to install app when .apk is downloaded
+                                                BroadcastReceiver onComplete = new BroadcastReceiver() {
+                                                    public void onReceive(Context ctxt, Intent intent) {
+                                                        ContextCompat.checkSelfPermission(ctxt, Manifest.permission.READ_EXTERNAL_STORAGE);
+
+                                                        installApk();
+                                                        getContext().unregisterReceiver(this);
+                                                    }
+                                                };
```
Simpler code not using string concatenation and converting from Uri to File to Uri (including the extra line after the BroadcastReceiver to have a blank line before registering the receiver):
```suggestion
                                                ContextCompat.checkSelfPermission(c, Manifest.permission.WRITE_EXTERNAL_STORAGE);
                                                File file = new File(c.getExternalFilesDir(null), "odysee-android.apk");

                                                if (file.exists()) {
                                                    file.delete();
                                                }

                                                DownloadManager.Request request = new DownloadManager.Request(Uri.parse(latestUrl));
                                                request.setTitle(getResources().getString(R.string.downloadmanager_downloading_new_version));

                                                //set destination
                                                request.setDestinationUri(Uri.fromFile(file));
                                                request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE);

                                                // get download service and enqueue file
                                                final DownloadManager manager = (DownloadManager) c.getSystemService(Context.DOWNLOAD_SERVICE);
                                                manager.enqueue(request);

                                                //set BroadcastReceiver to install app when .apk is downloaded
                                                BroadcastReceiver onComplete = new BroadcastReceiver() {
                                                    public void onReceive(Context context, Intent intent) {
                                                        ContextCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE);

                                                        installApk(file);
                                                        getContext().unregisterReceiver(this);
                                                    }
                                                };

```

kekkyojin replied:
Using a string to hold the path makes it easier to debug any problem in the future.

I will add the extra line.

ktprograms replied:
> Using a string to hold the path makes it easier to debug any problem in the future.

Why? You can still print `file.getAbsolutePath()`

ktprograms replied:
@kekkyojin ^

kekkyojin replied:
Printing some string value into logcat is not debugging. When debugging, you would need to see at every moment what value is stored on any variable.

ktprograms replied:
You can still debug the `file` variable. I don't see the need to have a string that gets converted to a `File`.

In Android Studio you can enter custom expressions in the debugger, so you can watch the value of `file.getAbsolutePath()`.

ktprograms reviewed (changes requested):
Works well, but I think a few code changes would be good.

Please try to get it working on the foss build also (should just be putting the fileprovider in `main/AndroidManifest.xml`)

Can the APK also be deleted after updating?

kekkyojin commented on app/src/main/java/com/odysee/app/ui/findcontent/AllContentFragment.java:
```diff
@@ -466,6 +496,172 @@ public void onPause() {
+                    InstallSourceInfo installSourceInfo = pm.getInstallSourceInfo(activity.getPackageName());
+                    installer = installSourceInfo.getInstallingPackageName();
+                } else {
+                    //noinspection deprecation
```
_Android Linter is wrong with this warning. Removing the noinspection keeps the deprecation notice

ktprograms replied:
I see.

kekkyojin committed:
DownloadManager wording for title

kekkyojin committed:
Move FileProvider to main flavor

kekkyojin committed:
Make snackbar dismiss-able

kekkyojin committed:
Use appropiate JSON key

kekkyojin commented on app/src/main/java/com/odysee/app/ui/findcontent/AllContentFragment.java:
```diff
@@ -466,6 +496,172 @@ public void onPause() {
+                        JSONObject responseJson = new JSONObject(respBody);
+
+                        long latestCode = responseJson.getLong("latestVersionCode");
+                        String latestUrl = responseJson.getString("url");
```
No. There is only one URL for all version codes, but version code could change. The URL will always point to the latest one. The version code on the JSON file is expected to always be the latest one, but naming it latestVersionCode makes it explicit.

ktprograms replied:
How will the url always point to the latest?

kekkyojin committed:
Applying some suggestions

tzarebczan committed:
Merge pull request #357 from OdyseeTeam/issue-template-correct-chat

issue_template: use correct chat link

ktprograms reviewed (commented):
What do these changes to the layout file do? (In https://github.com/OdyseeTeam/odysee-android/pull/355/commits/f8c16ee47e3e0097d9021aa4e2efba7ca4f02313)

kekkyojin committed:
Refactor intent to install APK file

kekkyojin commented on app/src/main/java/com/odysee/app/ui/findcontent/AllContentFragment.java:
```diff
@@ -466,6 +496,172 @@ public void onPause() {
+         *   url: string including https and package file }
+         */
+        Request.Builder builder = new Request.Builder();
+        builder = builder.url("http://apk.odysee.com/odysee-android/odysee-android-latest.json");
```
I used a different domain because maybe they would prefer to use it in order to keep it separate from the non-native release. Anyway, I was also expecting to put there any server pointed at by the team.

kekkyojin commented:
### How is this expected to work?

After a new version is released -I mean, after `version code` has been increased-, JSON file must be updated with new `version code` and the new APK file published on the URL. The URL is not expected to change between releases.

ktprograms commented:
I see, that makes sense.

ktprograms committed:
player: don't set newPlayerCreated when playerManager initialized

When newPlayerCreated is set to true, streaming url is refetched and
playback is restarted.

Also call setPlayer(null) in setPlayerForPlayerView. Without this
playback freezes when returning from picture in picture.

Fix: #356

ktprograms committed:
claim: use name if title is null in getTitleOrName

kekkyojin committed:
Set newly created channel as default one if it is the only one (#360)

kekkyojin committed:
Use odysee.tv for JSON file hosting

akinwale committed:
Playlists parity (#362)

* add description field for playlist publish

* allow reordering of playlist items with drag and drop

* delete from playlist, set / update language when publishing, fix crash errors

akinwale committed:
fix string value format

akinwale committed:
Advanced playlists parity (#366)

* create playlist overlay layout

* handle playlist playback using playlist overlay

* tweak playlist overlay background

* opening any related content item should clear the current playlist

* loop and shuffle playlists

* reset playlist overlay properly

* unify URLs being saved to the view history

* now playing queue

kekkyojin committed:
Simplify an if statement (#367)

kekkyojin committed:
Refactor SaveSharedUserState into a callable #364

kekkyojin committed:
Replace deprecated PlayerView with StyledPlayerView (#363)

kekkyojin committed:
Make playback quality reactive to current network connection (#373)

kekkyojin committed:
Unlock screen orientation for Android 12+ devices (#368)

ktprograms committed:
file view: remove not loading reactions on livestreams

ktprograms committed:
file view: align follow heart icon with Follow text

ktprograms committed:
file view: don't hide global player

Instead only hide it when playing another video/viewing content.

Update mini player bottom margins when starting file view fragment.

akinwale committed:
set the user's email address as the account username in AccountManager

akinwale committed:
bumpversion 0.0.16 --> 0.1.0

ktprograms committed:
file view: implement add to list, repost, download

- Use horizontal scroll view for file actions.
- Implement Add to List, Repost and Download.
- Pass auth token to repost task.
- Respect default channel in repost dialog.
- Use DownloadManager for downloading files.
- Change FilePickerListener#onFilePicked to receive original Uri data
  from the received intent. This is used for deleting the file that got
  created when selecting where to download the content. If the created
  file is not deleted, DownloadManager doesn't overwrite it, so there
  will be duplicated files, one of which is empty and invalid.
- Remove old code left behind from switching away from the on-device SDK
  which required downloading files before viewing.
- Skip initializing UI and loading claim content in onResume if
  returning from file picker.

ktprograms committed:
claim edit: fix channel creator getting shown when fragment resumes

Always set channel from claim even after first load from claim data.

ktprograms committed:
publish: fix nothing shown in channel selector when there's no channels

Expected to show "Create a channel" and Anonymous, but they're not
because the ClaimListTask which later calls the method to populate with
placeholder claims fails.

Remove code for showing channel creator from enableChannelSpinner(),
it's already handled in the spinner onItemSelectedListener.

ktprograms committed:
blocked & muted view: fix listing blocked and blocking/unblocking

Just skip single channel if it has no blocked channels instead of
throwing an error and stopping checking on all channels.

Don't get signing_channel on claims to unblock/unmute as only channels
can be blocked/muted so there's no file claims in the adapter.

Fix inverted and broken logic in context menu actions.

ktprograms committed:
claim context menu: add Report action, implement handlers on all screens

Currently, although all context menu actions are shown when opening the
context menu on claims on any screen, for most screens the actions don't
do anything. Implement the functionality for all screens.

In ClaimListAdapter, remove the longClickForContextMenu flag and always
set currentPosition to the item's absolute position in the adapter when
a long press occurs. In the context menu listeners, use
getCurrentPosition() instead of getPosition() as the latter doesn't
return the correct value on a long press.

On the search page, check if a claim is a channel, and if so don't
attempt to get its signing channel when blocking/muting.

kekkyojin committed:
Show livestreams also for Rabbit Hole category (#386)

ktprograms committed:
go live form: fix crash when going back immediately after entering

Check if activity is not null before initializing ReplaysPagerAdapter.

Also only load replays when actually on replays tab.

ktprograms committed:
chromecast: remove app bar cast button and introductory overlay

There is no custom web receiver app so casting is only useful when
playing a video.

ktprograms committed:
blocked & muted: load blocked channels in parallel

akinwale committed:
updated version code -> 1000

akinwale committed:
check eu-google rule if the app was installed from the Play Store

akinwale reviewed (approved):
This looks good to me. Can merge when it's ready.

akinwale committed:
add granular storage permissions for Android 13 and request for READ_EXTERNAL_STORAGE

akinwale committed:
check is_google_limited flag

akinwale committed:
only show the Blocked & Muted menu option when signed in

akinwale committed:
request specifically for READ_MEDIA_* permissions on Android 13

ktprograms committed:
file view: hide global player if continuing already playing claim

ktprograms committed:
chore(cleanup): remove DownloadActionListener

akinwale committed:
fix sign in UI thread crash errors on Android 13

ktprograms committed:
manifest: add <queries> for VIDEO_CAPTURE intent

Fix "no camera app available to record videos on this device" error.

akinwale committed:
Implement MANAGE_EXTERNAL_STORAGE permission for Android 13+

kekkyojin committed:
Smooth livestream's list displaying (#407)

* Show livestream's list with progress bar while loading content

* Don't show livestreams on categories content list

kekkyojin committed:
Merge branch 'master' into inapp-update

# Conflicts:
#	app/src/main/res/layout/fragment_all_content.xml

akinwale committed:
fix mute channel touch area layout

akinwale committed:
Mod screen (#412)

* blocked on behalf of layout updates

* moderated block screen and delegated moderation handling

kekkyojin committed:
Livechat tweaks (#411)

kekkyojin committed:
Unindent lines of code for updating reactions (#410)

kekkyojin committed:
Sort active livestreams on categories by descending viewers count (#409)

kekkyojin committed:
Make Android 7 the minimum required Android version (#405)

akinwale committed:
load the item tapped in the playlist overlay

kekkyojin committed:
Fix MANAGE_EXTERNAL_STORAGE permission's related crashes (#413)

kekkyojin committed:
Fix conflicts to merge with master

kekkyojin committed:
Remove branch for pre-Nougat devices

ktprograms committed:
file view player: auto exit fullscreen when switching to portrait

This reverts the orientation lock (preventing the player from leaving
fullscreen) added in
https://github.com/lbryio/lbry-android/commit/540a8412.

akinwale committed:
history screen

kekkyojin committed:
Replace some more newSingleThreadExecutor with OdyseeApp.getexecutor() (#417)

kekkyojin committed:
Add javadoc to global executor getters methods (#419)

kekkyojin committed:
Add some comments on build.gradle in relation to desugaring (#418)

kekkyojin committed:
Remove AccountManager permissions which are no longer needed (#425)

kekkyojin committed:
Show to which email address sign up verification message has been sent (#422)

kekkyojin committed:
Set text color to white on MaterialButtons (#423)

ktprograms committed:
file view: hide mini player and fix player after download picked

akinwale committed:
More playlist fixes (#428)

* improve navigation in publish playlist dialog

* fix autoplay and playback logic in a currently playing playlist

* display thumbnails for playlists on the Library screen

ktprograms committed:
claim_search: filter unlisted and future scheduled

Fix: #421

ktprograms committed:
file view: show unsupported content view, fix download toasts

If content is neither playable or viewable, show unsupported content
view with button to download the file.

Shut down scheduler used for observing download progress on error or
manual abort. This fixes error/success toasts being repeatedly shown.
(To get the success toast to repeatedly show, start a download, abort
it, then download again).

Also set downloadInProgress to false after download completes
successfully.

kekkyojin committed:
Change the way Play Services tasks are disabled (Linter suggestion) (#429)

* Change the way Play Services tasks are disabled (Linter suggestion)

* Change the way other Gradle tasks and configurations are modified

kekkyojin committed:
Merge from master and fix conflict

ktprograms committed:
manifest: set extractNativeLibs to true

Fix installation error due to zipalign not respecting page boundaries of
uncompressed .so files.

kekkyojin committed:
Send odysee.com as referer for media player requests

kekkyojin committed:
Add trailing slash to url

kekkyojin committed:
Fix crash when back button is pressed

tzarebczan committed:
Merge pull request #445 from OdyseeTeam/b444

Send odysee.com as referer for media player requests

tzarebczan committed:
Merge pull request #447 from OdyseeTeam/title-crash

Fix crash when back-button is pressed

kekkyojin committed:
Merge branch 'master' into inapp-update

kekkyojin committed:
Add new line to end of files
